### ISSUE TITLE
color(), material() and part()

### ISSUE THREAD

**Issue by doug-moen** [openscad/openscad#1608]
This feature request generalizes our export mechanism so that we can export models with:
* multiple named parts (which are allowed to overlap). For AMF, this means multiple &lt;object> elements. For DXF, and perhaps also SVG with Inkscape extensions, this means multiple layers. For STL and OFF, this means multiple output files are created by a single export.
* multiple colors and materials. The goal is to support 3D printing with multiple extruders, so colors are volumetric, rather than applied to the surface of a mesh. A single 3D part may have multiple non-overlapping volumes, each volume having a different color or material. A single 2D part may have multiple non-overlapping areas, each area having a different color. For AMF, this means each &lt;object>/part has multiple &lt;volume> elements. For DXF and SVG, each layer has multiple non-overlapping shapes, each of which can have a different solid color. For STL and OFF, this means multiple output files are created by a single export, one for each color/material.

The reason for bundling these three features together into one FR is to show how they interact with each another, and with the CSG operators. I also want to make the point that these 3 features are not special cases of some more general "metadata" feature. Different kinds of "metadata" interact with the CSG operators in different ways, depending on the role of that "metadata" within the system. For example, OpenSCAD2 has an "object" data type, which can be interpreted as a way to attach "metadata" to a shape. But the purpose and use of that metadata is different, and CSG operators must ignore OpenSCAD2 object attributes, while they must pay attention to the color(), material() and part() attributes. The CSG operators interact with the part() attribute differently than they interact with the color() and material() attributes.

`part("name")shape` associates the same part name with all of the components of `shape`, overriding any previously specified part names. If no part name is specified, the part name defaults to `""`. So `square(10)` is equivalent to `part("")square(10)`.

The concept of a 2D or 3D shape is generalized so that each shape has zero or more distinctly named parts, which may overlap. (As a special case, the empty OpenSCAD script produces the empty shape, which has no parts.) The CSG operations are generalized so that each distinct named part contained within the children list is operated on separately, and thus retains its identity.

For example,
```
union {
   part("p1") cube(10);
   part("p1") sphere(5);
   part("p2") cylinder(h=15,r=2);
}
```
This produces a new shape containing two parts named "p1" and "p2". The cube and sphere are unioned together into a single composite shape, since they belong to the same part. However, the cylinder is not unioned with anything, it retains its identity since it belongs to a separate part.

As a future extension, we'll be able to import an AMF file containing multiple &lt;object> elements, and preserve the identity of each &lt;object> as a separate part. Likewise for importing a DXF file with multiple layers.

A part name is zero or more characters chosen from the set [a-zA-Z0-9_]. That's intended to be a lowest common denominator design that will be compatible with all export formats.

The color() operator already exists, but now it is extended so that it affects the F6 rendering semantics. Shapes are generalized so that each part contains one or more disjoint regions, each region having a different colour. The CSG operators are generalized so that, when an operator is passed two shape arguments within the same part that have different colours, then the result is computed as some sort of composite, usually containing multiple regions with different colours. But these semantics vary from one operation to the next.

Colours are volumetric, and are considered to apply to each geometric point within a region, both at the surface and within the interior. If no colour is specified, then it defaults to `undef`. For example, `cube(10)` is equivalent to `color(undef)cube(10)`. You can remove all colours from a shape using `color(undef)shape`. This distinction matters for export: if a region has no colour, then no colour is specified for that region in the exported file.

For example, the `union` operator is extended so that the order of the shape arguments matters. If two shape arguments overlap, and belong to the same part, but have different colours, then the colour of the shape closest to the beginning of the argument list takes precedence for the colour of the overlapping region in the final result.

The material() attribute is far less important than part() or color(). I'm only suggesting it because the AMF file format allows you to specify a colour, a material, or both, for each volume in an object. The material attribute doesn't provide any benefit for any other currently supported export file format. We could leave it out without losing much. I'll update this paragraph later based on the discussion.

**Comment by kintel** [comment]
Thanks for the write-up doug - I'm interested in moving this forward and this is a good first step.
A few comments:
* multi-body STL (ASCII only) appear to be reasonably well supported by external software. We should consider doing the same, as a user-selectable export option. Similarly for the future extension of preserving part names on import.

* Have you considered alternative management of hierarchical naming? e.g.
```
MyModule() {
  part("p1") cube();
  part("p2") sphere();
}
part("ObjectA") translate([-10,0,0]) MyModule();
part("ObjectB") translate([10,0,0]) MyModule();
```

* This has some similarities to marker nodes (#952). Should we extend the .csg file format to include part names?

**Comment by doug-moen** [comment]
@kintel:

Part names are included in the CSG tree: they must be, since they change the semantics of CSG operations. Part names should appear in the *.csg output.

However, marker nodes (#952) are different from part names. The key is that marker nodes have no effect on CSG semantics. Marker nodes satisfy the same use case as OpenSCAD2 object attributes. They permit model- and application-specific annotations to be added to the node of a model: annotations which have no meaning to built-in operations. For example, a wheel component could have metadata giving the wheel radius, the width, the rim thickness, the number of spokes.

Hierarchical part names sound cool, but it's an extra complication that needs to be justified by use cases. It's something that could be added later: I've rigidly restricted the part namespace; a hierarchy separator character could be added later as an extension. Also, we'd need to look at each import and export format and see how hierarchical part names map in each case.

**Comment by t-paul** [comment]
Why does the actual use of the meta data need to influence the syntax? All meta data will need to be in the CSG tree anyway. Some would be only exported like the markers or BOM information and other can be handled either by the core (color), or the GUI (parameters, ...). In all other cases we try to move into the direction of unifying things, it seems strange not to do that for the meta data.

**Comment by doug-moen** [comment]
@t-paul: I don't really understand your comment "Why does the actual use of the meta data need to influence the syntax". What syntax? And why do you think that my proposal fails to move in the direction of unifying things? After all, I've unified #1041 and #1044 into a single more general concept of "part".

**Comment by t-paul** [comment]
Using the `part()` syntax is only possible for module instantiations, not for variable, function or modules definitions. At least I don't see how that would be possible.

**Comment by doug-moen** [comment]
I don't agree that the proposed color() and part() operators have anything to do with metadata. [Maybe you could claim that the *old* color() operator is metadata, since it has no effect on the model, only on the preview. But the *new* color() operator proposed here isn't metadata.]

The AMF file format has support for metadata via a &lt;metadata&gt; element. You can specify a textual description, author, copyright information, etc. This is clearly metadata, since it is information *about* the data, and it has no semantic effect on the model. It has no effect on what actually gets manufactured.

Similarly, if we can tag a module definition with a human readable explanation of what the module does, that is also metadata, since it has no effect on the operation of the module.

But in this proposal, color() and part() convey semantic information that is part of the model. They directly affect what gets manufactured. They aren't metadata. A corollary is that they do need to be included in the *.csg file.

**Comment by doug-moen** [comment]
@t-paul: The part() operator is a module; it takes a shape as an argument, and returns another shape. Ditto for color(). Semantically, they have the same domain and range as other modules like translate().

**Comment by doug-moen** [comment]
t-paul: Are you suggesting that the part() operator should be applicable to variable, function or modules definitions? If so, this makes no sense to me. In my mind, part() is a geometric operation. It sounds like you may have a counterproposal that is radically different from mine.

**Comment by kintel** [comment]
Some thoughts on the confusion of model attributes vs. metadata. I'm not sure if this really addresses the confusion/comments, but it's an attempt:

Essentially,  ```part```, ```color``` and ```material``` are all used to attach some sort of attribute to a node in the CSG tree. This proposal suggests some semantics for how this affects the underlying geometry, possibly hierarchically.

Now, we _could_ view ```part```, ```color``` and ```material``` as pure attributes, and also allow other metadata attributes. We then define that some attribute names have built-in behavior which affects the design in various ways. As a side note: One could argue that all transformation nodes are really just attributes as well. This could be useful in the future if we want to associate transformations with objects in a GUI.

This leads me to two issues:
* How do we attach such other metadata attributes to CSG tree nodes?
* How do we attach attributes to non-CSG tree nodes (e.g. modules, functions, variables)?

I think @t-paul's comments relate to these two issues.

..so the question to @doug-moen would be: Can we (later on) solve those two issues in a clean way, allowing us to move forward without doing it all at the same time?

**Comment by doug-moen** [comment]
@kintel: The proposed API for the part(), color() and material() modules is a pure functional interface that is completely abstract. This means the interface is independent of the underlying implementation. The underlying implementation could change in the future, and we wouldn't be forced to change the API or cause old code to stop working.

As soon as we extend OpenSCAD to allow arbitrary attributes (name/value pairs) to be attached to any entity, then the problem of data abstraction arises. There is now a distinction between stored and computed attributes. "Stored" attributes are stored directly in the data structure as name/value pairs, while "computed" attributes are computed from the values of other attributes when they are requested. We could decide to expose the underlying implementation, by requiring the user to use different syntax for stored vs computed attributes. Or we could hide the underlying implementation by providing an abstract, pure functional interface for accessing the attributes.

**Comment by kintel** [comment]
@doug-moen I don't quite follow your on what a "computed attribute" is..

**Comment by t-paul** [comment]
Yes, I'm not talking about the behavior, e.g. how color will be handled by the CSG operations.

In my view the color is a property of a specific node / mesh, and has no node identity in itself. Using `part()` and the already existing `color()` it create a new node and applies it's properties to the child nodes (that's what's visible to the user writing the code, yes we could hide this internally, but that's not my point). I'd prefer to see in the code and enforced by syntax that the information is only attached to a node, just like the '#' modifier right now.

The second point is currently coming from the Customizer topic which wants to annotate variables. I don't know if it's possible to handle `param(desc = "box width", values = [1:6]) size = 3;`. Also having no way to directly distinguish meta data from other things makes it more complicated to read and understand the code.

Also I think all meta data must go into the CSG file, that's the whole points of most of the meta data attributes (e.g. for the marker nodes). We will handle some of it internally, e.g. the GUI would handle the export related values. Basically every processor will just ignore all the attributes it does not know.

**Comment by kintel** [comment]
@t-paul Just a small comment on modifier characters: My current view is that this is syntactical sugar for setting a modifier attribute on a node. We could offer a full syntax for the same if we offer a general purpose attribute module.

**Comment by kintel** [comment]
Since customization UI is a much larger topic, I'd like to explore if we could decouple this proposal from the customization parameter annotations and let these concepts peacefully co-exist. In the future, they could potentially use the same underlying mechanism, or even use the same language front-end.

I think moving all these features forward in sync is a bit utopic with the current resource situation. ..so if we can see a clean break I think we should go for it.

**Comment by t-paul** [comment]
Yes, the modifier `#` could be a short-cut, but what I mean is, it's attached to the node, which is why it's not possible to write `# { cube(); sphere(); }` it needs a node to attach to, e.g. `union` and becomes an attribute of that node.

**Comment by doug-moen** [comment]
@t-paul wrote "Yes, the modifier # could be a short-cut, but what I mean is, it's attached to the node, which is why it's not possible to write # { cube(); sphere(); } it needs a node to attach to, e.g. union and becomes an attribute of that node."

See, this is an example of something that I consider a bug in OpenSCAD. As a user, I don't care about what a "node" is in the underlying C++ implementation. I just want things to work. The fact that this doesn't work because "it's not a node" is not something I want to hear. This is something that I had proposed to fix in the OpenSCAD2 design document.

Similarly, I don't want to be forced to care about how colours and parts are implemented internally.

I prefer the language to be simple, general, high level, abstract, and not shove implementation details into my face when it's not necessary.

**Comment by kintel** [comment]
@t-paul ah, ok, _now_ I understand what you meant an hour ago.

**Comment by doug-moen** [comment]
@t-paul wrote "Yes, the modifier # could be a short-cut, but what I mean is, it's attached to the node, which is why it's not possible to write # { cube(); sphere(); } it needs a node to attach to, e.g. union and becomes an attribute of that node."

Right now, I can write
```
color("red") {square(10);circle(5);}
```
and it works.

From your comments about color(), "I'd prefer to see in the code and enforced by syntax that the information is only attached to a node, just like the '#' modifier right now.", it seems you would like to introduce a new syntax for color in which the above code will no longer work.

**Comment by kintel** [comment]
I tend to agree with @doug-moen. Just because ```part``` doesn't create a traditional CSG node doesn't mean that it behaves differently than a CSG node from a user's perspective. The only place a user gets a whiff of what a "node" is, is when we save to a .csg file.

You can think of it as creating a "part node" and add the underlying nodes as children (this might very well also be the cleanest way of implementing it in the current codebase).

This is an implementation detail, until we write to file. When writing to file we need to transform our "node" concept into something compatible with each file format.

**Comment by t-paul** [comment]
Well, I guess we will just not agree on what's easier to read. That's not a big deal.

Back to the main question "how do I do that for module definitions and variables"? Having an annotation style syntax like a number of newer languages provide seems both useful and works for all cases I see right now.

**Comment by doug-moen** [comment]
The original implementation of `color()` did not support an alpha channel, and did not support symbolic color names. So the original interface was `color([r,g,b])`. When we upgraded the language to support those things, we extended `color()` so that you can write `color("red")`, `color("red",0.5)` and `color([r,g,b,a])`. The original interface still works for backward compatibility.

If we stop using an abstract, functional interface for specifying colour, and use "attribute" syntax instead, then does it become impossible to enhance the interface in a backward compatible way, as we did for the `color` module?

**Comment by kintel** [comment]
@doug-moen I my mind, attribute specification is still strongly rooted in the current syntax, making these two equivalent (note: attribute syntax popped off the top of my head for the sake of the argument):

```
   color([r,g,b]) someShape();
```

```
   attribute('color', [r,g,b]) someShape();
```

-> does that make sense?

@t-paul:

**When** we have a fully featured attribute syntax, including the proposed way of attaching any user-defined object dict ("POJO") to a shape, I think we would be close to satisfying @t-paul's request for a Customizer UI syntax. To make that complete, we need a way of attaching attributes to non-shape objects, which shouldn't be a problem as these attributes are really only currying the underlying value. We might, in addition, need some syntactical sugar for making this all more readable (e.g. the proposed javadoc-style syntax).

See a few messages above: If we can agree that this feature is orthogonal (for now) to the Customizer UI stuff, that would make it a lot easier to push this forward.

**Comment by bobc** [comment]
FWIW, I am entirely behind @doug-moen 's proposal. In my experimental Carve-CSG fork I implemented the same semantics, and I believe is the only way it can work.

It is important to realise that once you go to a volumetric representation, _the color/material is as fundamental part of the node as the geometry_. The two cannot be separated, therefore it is not possible to view it as an attribute. Unless you also regard the geometry as merely another attribute.

It is easy to get lost in semantics, but unless you can get over the mental hurdle of what volumetric semantics implies then no progress can be made.

One slight quibble, I think that AMF specifies non-overlapping volumes. In practice, this may not be too much of a problem, because slicers already handle overlapping STL files by using a priority scheme. So unless they are very strict, they could/are likely to apply priority scheme to AMF objects as well.

Otherwise, I'm fully with it.

**Comment by doug-moen** [comment]
Thanks, Bob, I based the color() proposal on your work. Maybe your fork can
be used to create the color() implementation.

In AMF, there is a two level hierarchy. There are multiple named objects
that can overlap. Within an object, there are multiple non-overlapping
volumes.

>

**Comment by guowcas** [comment]
@doug-moen : it looks like an assembly tree to me.

**Comment by doug-moen** [comment]
@guowcas: it's not an assembly tree, because there isn't a tree. The parts
are disjoint, they are not related by parent/child relationships. Maybe
what's confusing is my choice of the word "part" to describe the partition
of a model into disjoint, non-overlapping pieces which can be independently
exported.

An assembly tree would be a nice feature to have, but it's an entirely
separate feature from the "parts" feature. The closest we have to an
assembly tree right now is the CSG tree, but you can't annotate it with
high level information.

**Comment by guowcas** [comment]
Sure it's not an assembly tree yet, but maybe a tree like structure should be the final form of this evolution. However, I agree that a loose collection of parts is suitable for 3D printing by now since high level information is not required yet. 

best regards, 
from GuoW's iPhone

> 在 2017年4月24日，20:45，Doug Moen <notifications@github.com> 写道：
> 
> @guowcas: it's not an assembly tree, because there isn't a tree. The parts
> are disjoint, they are not related by parent/child relationships. Maybe
> what's confusing is my choice of the word "part" to describe the partition
> of a model into disjoint, non-overlapping pieces which can be independently
> exported.
> 
> An assembly tree would be a nice feature to have, but it's an entirely
> separate feature from the "parts" feature. The closest we have to an
> assembly tree right now is the CSG tree, but you can't annotate it with
> high level information.
>

**Comment by alromh87** [comment]
Hello, any help needed here?

**Comment by t-paul** [comment]
Yes, there's a number of topics to solve to get this feature in. The 2 main parts would be the ability to pass along color information inside the geometry engine and defining how the language extension can look like.

**Comment by traverseda** [comment]
I wrote a short script that lets you export multiple parts, figured this is the issue to mention it in as anyone searching for a way to do this will find it. It doesn't handle basically any of these edge cases and is only the most simple example, but it solved a problem for me.

http://localhost:8000/code/partsScad/index.html

**Comment by tcurdt** [comment]
@traverseda that's a link to `localhost`

**Comment by traverseda** [comment]
@tcurdt I have no idea what you're talking about :p

Thanks.

**Comment by cathalgarvey** [comment]
This issue is tracking two things, but are they closely related in implementation? I'm just wondering about colour support - I discovered today while hacking this little thing up:
https://git.hackers.town/seachaint/sheetstruder
..that colours are not supported in the output STLs. It looks like this is a long-standing thing without a clear priority, so I'm just wondering should I expect this limitation of OpenSCAD to change or not. :)

I don't actually have a multi-extruder printer in any case, but it would be nice if I could export models for gamedev or something, if I ever got the time to actually get into that.

**Comment by doug-moen** [comment]
@cathalgarvey The initial description describes very clearly why the `colour` and `part` are closely related in implementation. Just read it, please.

> it would be nice if I could export models for gamedev or something.
But in game dev, coloured models usually have photographic images texture mapped onto the triangles in the mesh. Or, alternatively, some mesh file formats like X3D allow you to specify a colour for each face, or a colour for each vertex. This face colouring is quite different from the volumetric colour model being proposed here, where each volumetric part in an assembly has a single colour attached to it.

This proposal is designed for multi-extruder FDM 3D printing, which has very different requirements than gamedev.

**Comment by gingerbeardman** [comment]
+1 for material

my use case would be to mark certain objects as reflective so they don't go black at some angles (given fixed lighting direction see #3339)

**Comment by nophead** [comment]
Or too bright at some angles. Threads look too bright compared to their base material because they face the light.

**Comment by gingerbeardman** [comment]
In my situation I don't mind if they blow out to white because after rendering I convert to black and white anyway. I just want all my whites to stay white, and not become some level of gray that will sometimes be crushed to black.

**Comment by KeithSloan** [comment]
+1 for material. I would like to convert OpenSCAD shapes to GDML solids. The only real thing stopping this is the lack of a material name that would be in the CSG definition.

This is a very old requirement request like five years at least, what is the likelyhood of an implementation.

Happy to donate if material name is addressed.

**Comment by traverseda** [comment]
@KeithSloan I was using my code above to convert stuff to GLTF, although that part of the code was never quite completed.

Here's the blender code I used to convert the files to GLTF, although it currently doesn't deal with multi-parts.

```
import sys
argv = sys.argv
infile, outfile = argv[argv.index("--") + 1:]

import bpy
bpy.ops.wm.read_factory_settings(use_empty=True)
bpy.ops.import_mesh.stl(filepath=infile)
material = bpy.data.materials.new(name="base")
material.use_nodes = True
principled_bsdf = material.node_tree.nodes['Principled BSDF']
principled_bsdf.inputs[0].default_value = (0.01, 0.2, 0.0, 1.0)
for o in bpy.context.editable_objects:
    o.data.materials.append(material)
bpy.ops.export_scene.gltf(filepath=outfile)
```

The problem with my approach is that it's slow, it needs to invoke a new instance of blender for every single part. It's always going to have that limitation and I'm not inclined to develop it further because it is a terrible hack happening at the wrong level in openscad.

Still, maybe the general approach will work for your use case? Let me know if you have any questions about it.

**Comment by KeithSloan** [comment]
Thanks but it does not work for me. If material was added to the scripting language, then I would use FreeCAD OpenSCAD importer to import the SCAD or CSG file. The FreeCAD OpenSCAD workbench runs the code to produce the CSG file under the covers. The plan would be to then use FreeCAD GDML workbench to export the model to GDML. Trouble is the objects need a GDML material allocated and this currently would have to be a manual process through the Gui, where as if one could specify a material when scripting things would be a lot easier. What is GDML used for? Its used for  Montecarlo simulation using  applications such as Geant4. Medical applications, satellites and stuff that goes into space need to be modelled to make sure sensitive parts are protected in a space environment. Some people design such stuff with CAD and export to Tessellated Objects but modelling then takes a long time where as people could script with OpenSCAD. GDML has a richer set of Computer Solids than OpenSCAD so it should not be a problem converting. Hull & Minkowski are a challenge but rich models can still be created without them or they could be converted to Tessellated. One of the guy's that starred the GDML repro was a MSc student designing a CubeSat. Now would it not be cool to have CubSat's and other stuff going to space, designed with OpenSCAD.

**Comment by pca006132** [comment]
Note that manifold supports vertex properties which are preserved across boolean operations. This is used in the [web demo](https://manifoldcad.org/) for color example (Tetrahedron puzzle):
![image](https://github.com/openscad/openscad/assets/12198657/47357545-4514-4dee-abf7-364e9662e62f)

We also support finding the original mesh id of different faces via mesh relation.

There are some issues (https://github.com/elalish/manifold/issues/536) regarding vertProperties in manifold because it is less tested, but we can easily fix them. I can perhaps try to integrate it in OpenSCAD...

**Comment by t-paul** [comment]
> We also support finding the original mesh id of different faces via mesh relation.

So this could also be used for viewport-selection -> select in code editor, like we have for the preview. That would be very nice.

**Comment by jsteinhofff** [comment]
> I wrote a short script that lets you export multiple parts, figured this is the issue to mention it in as anyone searching for a way to do this will find it. It doesn't handle basically any of these edge cases and is only the most simple example, but it solved a problem for me.
> 
> https://traverseda.github.io/code/partsScad/index.html

very easy and gets the job done, thanks for sharing @traverseda !

**Comment by gsohler** [comment]
Or you could approach a single-program solution
[image: image.png]

This is openscad with builtin python, and you get a 3mf file with both
parts included as a bonus
I am still happy to share

**Comment by jsteinhofff** [comment]
The image is broken for me, but I assume you pointed to pythonscad, which I just discovered in parallel coming across a reddit post from you. Love it, will give it a try asap, I am a big fan of python anyway!

**Comment by gsohler** [comment]
sorry for that. let me try it better

try this link:
https://imgur.com/a/heUN84M

Code is:

from openscad import *

rube=cube([10,10,10]).color("red")
blegg=sphere(10).resize([10,10,20]).translate(
    [5,5,10]).color("blue")

export(
    {"rube":rube,"blegg":blegg},
    "combined.3mf")

show([rube, blegg])


cheers

**Comment by jsteinhofff** [comment]
finally i found some time to look into this... lots of chaning the order of lines, few less curly brackets, few more comma ... it works!
Amazing work @gsohler , this boosts OpenSCADs capabilities by a huge margin :star_struck:
ps: i have to mention that i crashed OpenSCAD few times along the way, i will think a bit about it and give notice in your fork

**Comment by rob-miller** [comment]
It seems like someone should reference this April, 2023 thesis here: [Multi-Material Support in OpenSCAD](https://tams.informatik.uni-hamburg.de/publications/2023/BSc_Nicolaus_Johannes_Leopold.pdf).

His implementation is at [https://github.com/nle1990/openscad-mm/](https://github.com/nle1990/openscad-mm/).

**Comment by t-paul** [comment]
> His implementation is at https://github.com/nle1990/openscad-mm/.

Oh, my that's not even a fork :cry:. I hope at least the analysis will be useful, it looks like quite some effort went into the discussion of behavior.

**Comment by jordanbrown0** [comment]
In #5785, @jimb15122 said:
> May I suggest using the otherwise useless (for stl export) COLOR tag as the basis for the split. the resulting stl files would append the color (possibly in hex format) to the base filename. UI wise, all that would be needed is a new "export stl by color" option.

The problem is that the rendering engine historically hasn't, and still doesn't quite, maintain volumetric information about color.

With the old CGAL engine, all color is lost.  With the new Manifold engine some color is retained, but (a) I don't believe it's volumetric - that is, the faces are colored, not the interior, and (b) based on a model I'm working on at the moment I don't think it's quite right yet.

The previewer shows color, but it cheats.

**Comment by jimb15122** [comment]
I am by no means familiar with openscad internals.  But recently, I have been working on a Blender/Python script that reads a csg, and populates a blender scene.  It uses a stack to push and pop colors (and transforms) as they go in and out of scope. Then as objects are encountered, they get assigned a material based on the current color on the stack.   My point is (npi) that color is simply a state, and does not need to be intertwined with geometry.

**Comment by pca006132** [comment]
I just don't understand why the geometry engine has to know the volumetric information. Mesh is not supposed to maintain this. Can't we just process each material individually? E.g. maintain a mesh for each material, as well as a final aggregated mesh so we can perform subtraction/intersection correctly.

**Comment by jordanbrown0** [comment]
```
color("red") cube(10, center=true);
color("green") cube(10);
```

There is a 5-unit cube that is inside both the red and green cubes.  What color is it?  On the display, it doesn't matter, but when we turn it into plastic what color is it?

```
difference() {
    union() {
        color("red") cube(10, center=true);
        color("green") cube(10);
    }
    rotate(-135) translate([-15,0,-10]) cube(30);
}
```
Now, with one half cut away, you can see that formerly-interior cube.  What color is it?  Again, what color of plastic gets laid down?

```
intersection() {
    color("red") cube(10);
    color("green") cube(10, center=true);
}
```

What color is the result?  What color plastic is laid down?

**Comment by jordanbrown0** [comment]
Another way to look at it is that when you do a boolean operation (notably union; the others are easier) between meshes of two colors, the result usually needs to be two disjoint meshes of different colors.  In general, if you union an object of N colors with an object of M different colors, you end up with N+M meshes, each of a different color.

(There are exceptions for cases where two of the meshes are the same, or where one mesh is completely enclosed, but if we answer the questions for the simple cases then I think those cases are answered automatically.)

The matter is especially concrete for 3D printing, but still applies if all you want to do is display.  If you build a multi-colored object, and then you cut away parts of it, or you drill through it, you need to know and display the colors of the new faces.  The current preview logic paints the revealed faces with the color of the negative object, but that seems more like an implementation artifact than like a desirable and understandable outcome.

**Comment by gingerbeardman** [comment]
If it helps, I don't turn any of my designs into plastic :) just PNGs

**Comment by jordanbrown0** [comment]
If all you produce is PNGs, then the existing preview logic is mostly enough.  (Except Z-fighting.)

**Comment by jordanbrown0** [comment]
Well, give or take the cut-away / drill-through questions.  The existing logic has an answer for those, but it's arguably not the rightest answer.

**Comment by jimb15122** [comment]
Jordan, your questions are very valid.  They are at the heart of why no progress is being made.  The "stateful color" approach however can answer the question, albeit in a way you may not prefer.  I view the differencing union and intersect operations as resulting in a single mesh. same way cube or extrudes create a single mesh.  (This view is probably a benefit of my ignorance). Even when nested, the outermost bool operation is viewed as creating a single mesh.  Therefore the meshes entire color is the color that was in play at the time the outermost bool operation was encountered.

**Comment by jimb15122** [comment]
implicit / lazy unions can be used to keep colors separate.

**Comment by burunduk3** [comment]
Last time I had something similar to do, instead of color I used conditions, something like
```scad
if (target == "part1")
  square(10);
if (target == "part2")
  difference() {
    square(10, center=true);
    square(10);
  }
```

Then I ran `openscad -Dtarget=...` with different values and combined the results with an external script.

It wasn't too straightforward, though. At some point I found myself parsing CSG tree just to get coordinates of circle's center.

**Comment by reconbot** [comment]
Exactly, I don't think we need a perfect answer. I think we need something to make this easier.

**Comment by jordanbrown0** [comment]
>[my questions] are at the heart of why no progress is being made. 

Yes.  answering them is critical, and hard.  Or, more precisely, implementing an answer is hard.

>I view the differencing union and intersect operations as resulting in a single mesh.

Difference and intersection are actually the more-or-less easy ones.  You can just use the color(s) of the first child, and ignore the color(s) of the later children. That seems straightforward and reasonable.

It's union that's hard, because somehow you have to get rid of the overlap from one part or the other.

(Hull and Minkowski are just bad for multicolor; let's ignore them until we figure out the others.)

> Therefore the meshes entire color is the color that was in play at the time the outermost bool operation was encountered.

What does "that was in play" mean here?  If you're unioning a red cube with a green cube, what color is "in play"?  If you mean the color that wraps around the union, discarding the colors of the two cubes, well, sure, but that's really restrictive.  (How would you do my cutaway example above?)

> implicit / lazy unions can be used to keep colors separate.

Maybe.  But that assumes that they were separate in the first place, which is a pretty harsh restriction to put on the model.  If you're willing to live with that restriction, I'm pretty sure that with today's OpenSCAD I could write wrappers around `color()` and the three Boolean operators that would let you extract each color, one at a time.

**Comment by jordanbrown0** [comment]
>instead of color I used conditions

Sure.  It's just really inconvenient.

**Comment by jordanbrown0** [comment]
@doug-moen , if you're still around:  what's the use case for parts having overlapping volumes?

---

Here's some thoughts on the semantics that I would expect.

I think that materials and color can be considered the same sort of thing.  For FDM printing, they *are* the same sort of thing - they are different selections of which filament to lay down.  For the discussion below I'll ignore the distinction.

- But:  face coloring / texturing is very different from volumetric coloring/texturing.  When coloring faces, it's entirely meaningful for a cube to have six different colors, while that's not at all meaningful for volumetric coloring.  Face coloring is useful for 3D models intended only for display, e.g. for games.

I think that 2D and 3D operations are basically the same; the only difference is in extrusion and projection.  I'll cover those below.

At all points while processing, where we currently have only a single homogeneous object, we need to have a list of sub-objects with different attributes.  (Note:  this distinction is visible only inside the rendering process; it is not visible in the CSG tree.)  Any Boolean operation b(A,B) takes as inputs two lists of sub-objects, and returns as output a list of sub-objects.  I believe that all of the Boolean operations that we support are left-associative - that is, b(A,B,C) is equivalent to b(b(A,B),c), and so I'll discuss only two children.

The two questions are what the rules are for each Boolean operation, and how those rules are implemented.

Taking the easiest first:  difference.  Only the first child of a difference is positive; the other children are negative.  Since negative objects do not appear in the final result they have no attributes; their attributes can be ignored.  For each sub-object in child A, subtract each of the sub-objects in child B, yielding a (possibly empty and discarded) sub-object in the result.  (Or, equivalently, union the sub-objects in child B, and subtract that union from each of the sub-objects in child A.)

Next:  intersection.  Although in theory intersection is commutative, we need to make a choice which object "wins".  It seems reasonable to have the first object "win", making intersection similar to difference:  there is an original object A, and the second object B says which parts of A to trim away.  Here you must union the sub-objects in B and discard their attributes; for each sub-object in A you intersect with the union of the sub-objects in B.

Arguably the hardest is union.  You have to figure out what to do with the overlap, while retaining the non-overlapped volumes with their original attributes.  Because we have otherwise had the first child "win", I would maintain that here:  the overlapping volume derives its attributes from child A.  The first step in the operation is trivial:  child A is unchanged, for the moment.  Next, subtract A from B (as defined above), yielding the parts of B that do not overlap A; call that B-prime.    Merge A and B-prime, unioning those sub-objects that have the same attributes.

If we address those three, I think we have a viable solution.  But we should continue to discuss the other operators...

Extrusion.  For both linear and rotational extrusion, I think you can separately extrude each sub-object in the 2D input, maintaining its attributes, yielding a list of sub-objects in the 3D output.

For projection, Minkowski sum, hull, offset, and fill, I would initially union all sub-objects and discard all attributes, yielding a single default-attribute object.  It might be possible to do something cleverer in the future.  It seems like future cleverness would be compatible.

Those schemes work reasonably from a pure mathematical perspective.  It's not clear, however, how practical they are.  They rely on unioning and subtracting meshes that may have been generated independently and may be mathematically identical, but may in practice be microscopically different.  It would be interesting to throw together a proof of concept to see whether there are practical issues.

**Comment by jimb15122** [comment]
re: union is most challenging 
grok: A union is:  The act or state of joining or uniting two or more things into one.

The operative word is "one".  Therefore, choosing a single color makes some sense.  in the following example, the "color in play" that i referred to earlier would be yellow.

color ("yellow") {
  union(){
    color("red") sphere(20);
    color("blue") cube([10,20,30]);
  }
}

there are many ways to interpret this.  The primary one being the author is asking for trouble.

but a possible one would be:
-  use yellow as the default color for child objects that have none.  if the red and blue color tags were omitted, the resulting unioned object would just be yellow. 
 - allow inner color tags to override the "in play" color.
 - average the rgba of all inner objects to get the unions effective color.  in this case, purple.

for simplicity, i would actually prefer to ignore all inner colors, and use yellow.  As that is what happens now if colors are specified for any of the 2d objects inside an extrude.  I was surprised but they were ignored. the result of an extrude can only be the color that was "in play" for the entire extrude.  

Re: face color vs solid color:  this concept is not currently well documented anywhere in openscad and would open a huge can of worms. (yes, i am aware the color of the drill bit affects difference() seems a bit odd. ).  import the csg into blender for that kind of stuff.  Therefore i see no problem using simple

once a decision is made, it should be straightforward to add "export stl by color". to generate multiple stl s for each color encountered with color name (as coded) appended to the base file name.

**Comment by jordanbrown0** [comment]
That example is easy, at least in my opinion.  The resulting object should be yellow.

It's much harder if you omit the `color("yellow")`, leaving only the `union()`.  There I would say that you should - must, really - have a red sphere and a blue cube.  Or actually one, and part of the other, so that they don't overlap.  If you say that the union applies, say, the color of the first child to the entire result, that makes working with multicolored objects just awful.  How would you define a module `red_sphere_and_blue_cube()`?  If you can't have a module return a multi-colored object, IMHO it's all a waste of time.  And if it *can* return a multi-colored object, by doing some sort of partial union, then `union()` could do the same.

**Comment by gingerbeardman** [comment]
I have a lot of respect for OSS.

It feels like this has been prematurely optimised or is wrapped in a lot of red tape.

**Comment by jimb15122** [comment]
Jordan,  Openscad is a wonderful tool and I am forever appreciative of its elegant design.  

Given: Problems with deciding what the color should be only occur in ambiguous situations.

Given: It is unthinkable to introduce a change that would cause big changes with how things look in the GUI.

Given: Humans have the ability to avoid ambiguity if properly educated in identifying the situations they arise.

Given: The ability to generate multiple parts in a single export can save much effort.

Given: Adding a new "part" tag will most likely suffer the same ambiguities. (unless well documented), but that would be nicer, as you could just assign an arbitrary part name, and it avoids religious wars over what color means.

Therefore:  It does not matter what color you choose in ambiguous situations, just document the current behavior.  

If you need this feature it will be much easier to avoid the ambiguities, than to implement logic to generate possibly hundreds of parts in separate executions.

**Comment by jordanbrown0** [comment]
I agree with most of what you say.

The problem is that for printing you need volumetric color - you need to know the color of a volume of space.  What we have now is face coloring - we know the colors of the faces that surround that volume.

There are several different ways that you can end up with *one* object with different colors on different faces.  There's nothing wrong with that, if all you want is a mesh (a bunch of connected triangles).  In fact, for some applications (notably gaming), that's exactly what you want.  But if you need to know what color (or material) to make the object out of, you need something different.

If you union overlapping red and blue cubes, I don't care much what color the overlap volume is - but it needs to have a color, and it needs to have internal boundaries between the two colors, and the behavior needs to be understandable and predictable.

**Comment by jimb15122** [comment]
All I need for printing (and multi-part output) is a way to keep various collections of volumes (meshes) separate, and assign each a name.  Using color for this purpose seemed logical at first as it would not have required an extension to the language.  But given the volume vs face internal complications (which have no bearing on the problem I am trying to solve) it is probably best to introduce a new feature.  

A new "part" element would have the same syntax as color(), but the string argument can be any name.  To avoid ambiguity, an exception or warning is thrown if it nested inside a parent part.  It is legal to use the same part name for geometry generated across multiple modules, but nesting is not permitted.  Geometry that is created outside the scope of a part() is given a default name.

A new export option would NOT be needed.  The use of a part() in any script would be the indication that multiple files should be generated with .{partname} appended to the base filename.

module table(){
  part("top"){
    make_top();
    make_skirt();
  } // of top
  
  part("legs") {
     for ( side = ["left","right"], edge=["front","back"] )
       make_leg( side, edge );
  } // of legs
} // of table

when exported as "table" (to any format) would create table.top.xxx and table.legs.xxx

**Comment by jimb15122** [comment]
re: attaching metadata to csg tree nodes.  

I needed to do this as part of a project that imports csg files into blender for high quality visualization.  I wanted to communicate specific blender settings like background images, color to material maps, etc.

The workaround I came up with was to use text(text=metadata, size=0);  
works like a charm.  shows up in the csg, but does not render in openscad.  the size 0 flags it as metadata in the import logic.

**Comment by t-paul** [comment]
> The workaround I came up with was to use text(text=metadata, size=0);

Using invalid values *will* haunt you at some point.

**Comment by jimb15122** [comment]
> > The workaround I came up with was to use text(text=metadata, size=0);
> 
> Using invalid values _will_ haunt you at some point.

Not sure zero is invalid.  How about .0000000001.   Regardless, given the delivery rate of official releases, I won't be loosing sleep waiting for a breaking change.

**Comment by jimb15122** [comment]
> And yes, the current face-coloring semantics are undocumented, somewhat intentionally, precisely because they were known to be deficient. I started to document them and was told not to, because we would likely want to change them to implement volumetric color.

Seems odd that a system that does not really support 2d design at all (using default extrusions) to display 2d objects, would implement color() as a 2d face feature.  What possible use does that have in a system with such limited display rendering capabilities, but very great "solid" modeling capabilities.  color() should be implemented as volumetric only to align with the rest of the systems strengths.  And document what happens when ambiguity arises.  (nested colors should be blended in my opinion, but that is a huge breaking change)

**Comment by jordanbrown0** [comment]
The historical answer is that the previewer is very very different from the renderer, and can present a reasonable color picture fairly simply.  (It just throws triangles at the graphics subsystem, and the graphics subsystem figures out what is in front and displays it.)  It just automatically works for union cases, and for difference cases I think it ends up being easier to have difference paint a new color on newly exposed surfaces.

It's only been very recently that the renderer has had any capability to retain color, and the 2D subsystem's renderer still doesn't.

And it doesn't surprise me that the renderer likes face coloring, because it's basically a mesh processor, *not* a volume processor.
